// @douyinfe/semi-ui/lib/cjs/form/group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _isString2 = _interopRequireDefault(require("lodash/isString"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("@douyinfe/semi-foundation/lib/cjs/form/utils");

var _constants = require("@douyinfe/semi-foundation/lib/cjs/form/constants");

var ObjectUtil = _interopRequireWildcard(require("@douyinfe/semi-foundation/lib/cjs/utils/object"));

var _errorMessage = _interopRequireDefault(require("./errorMessage"));

var _label = _interopRequireDefault(require("./label"));

var _context = require("./context");

var _index = require("./hooks/index");

var _inputGroup = _interopRequireDefault(require("../input/inputGroup"));

var _index2 = require("../grid/index");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable react/destructuring-assignment */


const prefix = _constants.cssClasses.PREFIX; // Group component to remove Labels and ErrorMessages from its child fields
// Unified insertion of Labels and ErrorMessages from the group level
// Get Errors of all field in this group

const GroupError = props => {
  const {
    fieldSet
  } = props;
  const formState = (0, _index.useFormState)();
  const error = fieldSet.map(field => ObjectUtil.get(formState.errors, field));

  if ((0, _utils.isValid)(error)) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_errorMessage.default, {
    error: error,
    showValidateIcon: props.showValidateIcon,
    isInInputGroup: props.isInInputGroup
  });
};

class FormInputGroup extends _react.Component {
  renderLabel(label, formProps) {
    if (label) {
      if ((0, _isString2.default)(label)) {
        return /*#__PURE__*/_react.default.createElement(_label.default, {
          width: formProps.labelWidth,
          text: label
        });
      } else {
        return /*#__PURE__*/_react.default.createElement(_label.default, Object.assign({
          width: formProps.labelWidth
        }, label));
      }
    }

    return null;
  }

  render() {
    const _a = this.props,
          {
      children,
      label
    } = _a,
          rest = __rest(_a, ["children", "label"]);

    const updater = this.context;
    const formProps = updater.getFormProps(['labelPosition', 'labelWidth', 'labelAlign', 'showValidateIcon', 'wrapperCol', 'labelCol']);
    const labelPosition = this.props.labelPosition || formProps.labelPosition;
    const groupFieldSet = [];

    const inner = _react.default.Children.map(children, child => {
      if (child && child.props && child.props.field) {
        groupFieldSet.push(child.props.field);
        return /*#__PURE__*/_react.default.cloneElement(child, {
          isInInputGroup: true // noErrorMessage: true,
          // noLabel: true

        });
      }

      return null;
    });

    const groupCls = (0, _classnames.default)({
      ["".concat(prefix, "-field-group")]: true
    });
    const labelCol = formProps.labelCol;
    const wrapperCol = formProps.wrapperCol;
    const labelAlign = formProps.labelAlign;
    const appendCol = labelCol && wrapperCol;
    const labelColCls = labelCol ? "".concat(prefix, "-col-").concat(labelAlign) : '';
    const labelContent = this.renderLabel(label, formProps);

    const inputGroupContent = /*#__PURE__*/_react.default.createElement(_inputGroup.default, Object.assign({}, rest), inner);

    const groupErrorContent = /*#__PURE__*/_react.default.createElement(GroupError, {
      fieldSet: groupFieldSet,
      showValidateIcon: formProps.showValidateIcon,
      isInInputGroup: true
    });

    let content;

    switch (true) {
      case !appendCol:
        content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, labelContent, /*#__PURE__*/_react.default.createElement("div", null, inputGroupContent, groupErrorContent));
        break;

      case appendCol && labelPosition === 'top':
        // When labelPosition is top, you need to add an overflow hidden div to the label, otherwise it will be arranged horizontally
        content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
          style: {
            overflow: 'hidden'
          }
        }, /*#__PURE__*/_react.default.createElement(_index2.Col, Object.assign({}, labelCol, {
          className: labelColCls
        }), labelContent)), /*#__PURE__*/_react.default.createElement(_index2.Col, Object.assign({}, wrapperCol), inputGroupContent, groupErrorContent));
        break;

      case appendCol && labelPosition !== 'top':
        content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index2.Col, Object.assign({}, labelCol, {
          className: labelColCls
        }), labelContent), /*#__PURE__*/_react.default.createElement(_index2.Col, Object.assign({}, wrapperCol), inputGroupContent, groupErrorContent));
        break;

      default:
        break;
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      "x-label-pos": labelPosition,
      className: groupCls
    }, content);
  }

}

FormInputGroup.contextType = _context.FormUpdaterContext;
var _default = FormInputGroup;
exports.default = _default;
