// @douyinfe/semi-ui/lib/cjs/descriptions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _isPlainObject2 = _interopRequireDefault(require("lodash/isPlainObject"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/descriptions/constants");

require("@douyinfe/semi-foundation/lib/cjs/descriptions/descriptions.css");

var _descriptionsContext = _interopRequireDefault(require("./descriptions-context"));

var _item = _interopRequireDefault(require("./item"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const prefixCls = _constants.cssClasses.PREFIX;

class Descriptions extends _react.PureComponent {
  render() {
    const {
      align,
      row,
      size,
      className,
      style,
      children,
      data
    } = this.props;
    const classNames = (0, _classnames.default)(prefixCls, className, {
      ["".concat(prefixCls, "-").concat(align)]: !row,
      ["".concat(prefixCls, "-double")]: row,
      ["".concat(prefixCls, "-double-").concat(size)]: row
    });
    const childrenList = data && data.length ? data.map((item, index) => (0, _isPlainObject2.default)(item) ? /*#__PURE__*/_react.default.createElement(_item.default, Object.assign({
      itemKey: item.key
    }, item, {
      key: index
    }), item.value) : null) : children;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classNames,
      style: style
    }, /*#__PURE__*/_react.default.createElement("table", null, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement(_descriptionsContext.default.Provider, {
      value: {
        align
      }
    }, childrenList))));
  }

}

Descriptions.Item = _item.default;
Descriptions.contextType = _descriptionsContext.default;
Descriptions.propTypes = {
  align: _propTypes.default.oneOf(_constants.strings.ALIGN_SET),
  row: _propTypes.default.bool,
  size: _propTypes.default.oneOf(_constants.strings.SIZE_SET),
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  data: _propTypes.default.arrayOf(_propTypes.default.shape({
    key: _propTypes.default.node,
    value: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
    hidden: _propTypes.default.bool,
    className: _propTypes.default.string,
    style: _propTypes.default.object
  }))
};
Descriptions.defaultProps = {
  align: 'center',
  row: false,
  size: 'medium',
  data: []
};
var _default = Descriptions;
exports.default = _default;
