// @douyinfe/semi-ui/lib/cjs/badge/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _isString2 = _interopRequireDefault(require("lodash/isString"));

var _isNumber2 = _interopRequireDefault(require("lodash/isNumber"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _context = _interopRequireDefault(require("../configProvider/context"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/badge/constants");

require("@douyinfe/semi-foundation/lib/cjs/badge/badge.css");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const prefixCls = _constants.cssClasses.PREFIX;

class Badge extends _react.PureComponent {
  render() {
    const {
      direction
    } = this.context; // DefaultPosition here, static can't get this

    const defaultPosition = direction === 'rtl' ? 'leftTop' : 'rightTop'; // eslint-disable-next-line max-len

    const _a = this.props,
          {
      count,
      dot,
      type,
      theme,
      position = defaultPosition,
      overflowCount,
      style,
      children,
      className
    } = _a,
          rest = __rest(_a, ["count", "dot", "type", "theme", "position", "overflowCount", "style", "children", "className"]);

    const custom = count && !((0, _isNumber2.default)(count) || (0, _isString2.default)(count));
    const showBadge = count !== null && typeof count !== 'undefined';
    const wrapper = (0, _classnames.default)(className, {
      ["".concat(prefixCls, "-").concat(type)]: !custom,
      ["".concat(prefixCls, "-").concat(theme)]: !custom,
      ["".concat(prefixCls, "-").concat(position)]: Boolean(position) && Boolean(children),
      ["".concat(prefixCls, "-block")]: !children,
      ["".concat(prefixCls, "-dot")]: dot,
      ["".concat(prefixCls, "-count")]: !dot && !custom && showBadge,
      ["".concat(prefixCls, "-custom")]: custom
    });
    let content;

    if ((0, _isNumber2.default)(count)) {
      content = overflowCount && overflowCount < count ? "".concat(overflowCount, "+") : "".concat(count);
    } else {
      content = count;
    }

    return /*#__PURE__*/_react.default.createElement("span", Object.assign({
      className: prefixCls
    }, rest), children, /*#__PURE__*/_react.default.createElement("span", {
      className: wrapper,
      style: style,
      "x-semi-prop": "count"
    }, dot ? null : content));
  }

}

exports.default = Badge;
Badge.contextType = _context.default;
Badge.propTypes = {
  count: _propTypes.default.node,
  dot: _propTypes.default.bool,
  type: _propTypes.default.oneOf(_constants.strings.TYPE_SET),
  theme: _propTypes.default.oneOf(_constants.strings.THEME_SET),
  position: _propTypes.default.oneOf(_constants.strings.POS_SET),
  overflowCount: _propTypes.default.number,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  onClick: _propTypes.default.func,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func
};
Badge.defaultProps = {
  dot: false,
  type: 'primary',
  theme: 'solid',
  className: '',
  onClick: () => _noop2.default,
  onMouseEnter: () => _noop2.default,
  onMouseLeave: () => _noop2.default
};
