// @douyinfe/semi-ui/lib/cjs/toast/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ToastFactory = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _toastListFoundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/toast/toastListFoundation"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/toast/constants");

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _toast = _interopRequireDefault(require("./toast"));

require("@douyinfe/semi-foundation/lib/cjs/toast/toast.css");

var _uuid = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/utils/uuid"));

var _useToast = _interopRequireDefault(require("./useToast"));

var _cssAnimation = _interopRequireDefault(require("../_cssAnimation"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-param-reassign */
const createBaseToast = () => {
  var _a;

  return _a = class ToastList extends _baseComponent.default {
    constructor(props) {
      super(props);
      this.state = {
        list: [],
        removedItems: [],
        updatedItems: []
      };
      this.foundation = new _toastListFoundation.default(this.adapter);
    }

    get adapter() {
      return Object.assign(Object.assign({}, super.adapter), {
        updateToast: (list, removedItems, updatedItems) => {
          this.setState({
            list,
            removedItems,
            updatedItems
          });
        }
      });
    }

    static create(opts) {
      var _a;

      const id = (_a = opts.id) !== null && _a !== void 0 ? _a : (0, _uuid.default)('toast'); // this.id = id;

      if (!ToastList.ref) {
        const div = document.createElement('div');

        if (!this.wrapperId) {
          this.wrapperId = (0, _uuid.default)('toast-wrapper').slice(0, 26);
        }

        div.className = _constants.cssClasses.WRAPPER;
        div.id = this.wrapperId;
        div.style.zIndex = String(typeof opts.zIndex === 'number' ? opts.zIndex : ToastList.defaultOpts.zIndex);
        ['top', 'left', 'bottom', 'right'].map(pos => {
          if (pos in ToastList.defaultOpts || pos in opts) {
            const val = opts[pos] ? opts[pos] : ToastList.defaultOpts[pos];
            div.style[pos] = typeof val === 'number' ? "".concat(val, "px") : val;
          }
        }); // document.body.appendChild(div);

        if (ToastList.defaultOpts.getPopupContainer) {
          const container = ToastList.defaultOpts.getPopupContainer();
          container.appendChild(div);
        } else {
          document.body.appendChild(div);
        }

        _reactDom.default.render( /*#__PURE__*/_react.default.createElement(ToastList, {
          ref: instance => ToastList.ref = instance
        }), div, () => {
          ToastList.ref.add(Object.assign(Object.assign({}, opts), {
            id
          }));
        });
      } else {
        const node = document.querySelector("#".concat(this.wrapperId));
        ['top', 'left', 'bottom', 'right'].map(pos => {
          if (pos in opts) {
            node.style[pos] = typeof opts[pos] === 'number' ? "".concat(opts[pos], "px") : opts[pos];
          }
        });

        if (ToastList.ref.has(id)) {
          ToastList.ref.update(id, Object.assign(Object.assign({}, opts), {
            id
          }));
        } else {
          ToastList.ref.add(Object.assign(Object.assign({}, opts), {
            id
          }));
        }
      }

      return id;
    }

    static close(id) {
      if (ToastList.ref) {
        ToastList.ref.remove(id);
      }
    }

    static destroyAll() {
      if (ToastList.ref) {
        ToastList.ref.destroyAll();
        const wrapper = document.querySelector("#".concat(this.wrapperId));

        _reactDom.default.unmountComponentAtNode(wrapper);

        wrapper && wrapper.parentNode.removeChild(wrapper);
        ToastList.ref = null;
        this.wrapperId = null;
      }
    }

    static getWrapperId() {
      return this.wrapperId;
    }

    static info(opts) {
      if (typeof opts === 'string') {
        opts = {
          content: opts
        };
      }

      return this.create(Object.assign(Object.assign(Object.assign({}, ToastList.defaultOpts), opts), {
        type: 'info'
      }));
    }

    static warning(opts) {
      if (typeof opts === 'string') {
        opts = {
          content: opts
        };
      }

      return this.create(Object.assign(Object.assign(Object.assign({}, ToastList.defaultOpts), opts), {
        type: 'warning'
      }));
    }

    static error(opts) {
      if (typeof opts === 'string') {
        opts = {
          content: opts
        };
      }

      return this.create(Object.assign(Object.assign(Object.assign({}, ToastList.defaultOpts), opts), {
        type: 'error'
      }));
    }

    static success(opts) {
      if (typeof opts === 'string') {
        opts = {
          content: opts
        };
      }

      return this.create(Object.assign(Object.assign(Object.assign({}, ToastList.defaultOpts), opts), {
        type: 'success'
      }));
    }

    static config(opts) {
      ['top', 'left', 'bottom', 'right'].forEach(pos => {
        if (pos in opts) {
          ToastList.defaultOpts[pos] = opts[pos];
        }
      });

      if (typeof opts.zIndex === 'number') {
        ToastList.defaultOpts.zIndex = opts.zIndex;
      }

      if (typeof opts.duration === 'number') {
        ToastList.defaultOpts.duration = opts.duration;
      }

      if (typeof opts.getPopupContainer === 'function') {
        ToastList.defaultOpts.getPopupContainer = opts.getPopupContainer;
      }
    }

    has(id) {
      return this.foundation.hasToast(id);
    }

    add(opts) {
      return this.foundation.addToast(opts);
    }

    update(id, opts) {
      return this.foundation.updateToast(id, opts);
    }

    remove(id) {
      return this.foundation.removeToast(id);
    }

    destroyAll() {
      return this.foundation.destroyAll();
    }

    render() {
      let {
        list
      } = this.state;
      const {
        removedItems,
        updatedItems
      } = this.state;
      list = Array.from(new Set([...list, ...removedItems]));
      const updatedIds = updatedItems.map(_ref => {
        let {
          id
        } = _ref;
        return id;
      });

      const refFn = toast => {
        var _a;

        if (((_a = toast === null || toast === void 0 ? void 0 : toast.foundation) === null || _a === void 0 ? void 0 : _a._id) && updatedIds.includes(toast.foundation._id)) {
          toast.foundation.setState({
            duration: toast.props.duration
          });
          toast.foundation.restartCloseTimer();
        }
      };

      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, list.map((item, index) => {
        const isRemoved = removedItems.find(removedItem => removedItem.id === item.id) !== undefined;
        return /*#__PURE__*/_react.default.createElement(_cssAnimation.default, {
          key: item.id,
          motion: item.motion,
          animationState: isRemoved ? "leave" : "enter",
          startClassName: isRemoved ? "".concat(_constants.cssClasses.PREFIX, "-animation-hide") : "".concat(_constants.cssClasses.PREFIX, "-animation-show")
        }, _ref2 => {
          let {
            animationClassName,
            animationEventsNeedBind,
            isAnimating
          } = _ref2;
          return isRemoved && !isAnimating ? null : /*#__PURE__*/_react.default.createElement(_toast.default, Object.assign({}, item, {
            className: (0, _classnames.default)({
              [item.className]: Boolean(item.className),
              [animationClassName]: true
            })
          }, animationEventsNeedBind, {
            style: Object.assign({}, item.style),
            close: id => this.remove(id),
            ref: refFn
          }));
        });
      }));
    }

  }, _a.defaultOpts = {
    motion: true,
    zIndex: 1010,
    content: ''
  }, _a.propTypes = {
    content: _propTypes.default.node,
    duration: _propTypes.default.number,
    onClose: _propTypes.default.func,
    icon: _propTypes.default.node,
    direction: _propTypes.default.oneOf(_constants.strings.directions)
  }, _a.defaultProps = {}, _a;
};

class ToastFactory {
  static create(config) {
    const newToast = createBaseToast();
    newToast.useToast = _useToast.default;
    config && newToast.config(config);
    return newToast;
  }

}

exports.ToastFactory = ToastFactory;

var _default = ToastFactory.create();

exports.default = _default;
