// @douyinfe/semi-ui/lib/cjs/card/meta.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/card/constants");

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const prefixcls = _constants.cssClasses.PREFIX;

class Meta extends _react.PureComponent {
  render() {
    const _a = this.props,
          {
      avatar,
      className,
      description,
      style,
      title
    } = _a,
          others = __rest(_a, ["avatar", "className", "description", "style", "title"]);

    const metaCls = (0, _classnames.default)("".concat(prefixcls, "-meta"), className);

    const avatarNode = avatar && /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixcls, "-meta-avatar")
    }, avatar);

    const titleNode = title && /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixcls, "-meta-wrapper-title")
    }, title);

    const descriptionNode = description && /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixcls, "-meta-wrapper-description")
    }, description);

    const wrapper = title || description ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixcls, "-meta-wrapper")
    }, titleNode, descriptionNode) : null;
    return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, others, {
      className: metaCls,
      style: style
    }), avatarNode, wrapper);
  }

}

Meta.propTypes = {
  avatar: _propTypes.default.node,
  className: _propTypes.default.string,
  description: _propTypes.default.node,
  style: _propTypes.default.object,
  title: _propTypes.default.node
};
var _default = Meta;
exports.default = _default;
