// @douyinfe/semi-ui/lib/cjs/calendar/timeCol.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _localeConsumer = _interopRequireDefault(require("../locale/localeConsumer"));

var _context = _interopRequireDefault(require("../locale/context"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/calendar/constants");

require("@douyinfe/semi-foundation/lib/cjs/calendar/calendar.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const prefixCls = "".concat(_constants.cssClasses.PREFIX, "-time");

class timeCol extends _react.PureComponent {
  formatTime(item) {
    const {
      renderTimeDisplay
    } = this.props;

    if (typeof renderTimeDisplay === 'function') {
      return renderTimeDisplay(item);
    } else {
      const replaceTime = (template, time) => template.replace('${time}', String(time));

      return /*#__PURE__*/_react.default.createElement(_localeConsumer.default, {
        componentName: "Calendar",
        key: "locale-".concat(item)
      }, locale => {
        let time = item < 12 ? replaceTime(locale.AM, item) : replaceTime(locale.PM, item - 12);

        if (item === 12) {
          time = replaceTime(locale.PM, item);
        }

        return time;
      });
    }
  }

  renderTime() {
    const {
      className
    } = this.props;
    const wrapperCls = (0, _classnames.default)(className, "".concat(prefixCls));
    const list = [...Array(24).keys()].map(item => this.formatTime(item));
    list.splice(0, 1, '');
    const inner = list.map((item, index) => /*#__PURE__*/_react.default.createElement("li", {
      key: "time-".concat(index),
      className: "".concat(prefixCls, "-item")
    }, /*#__PURE__*/_react.default.createElement("span", null, item)));
    return /*#__PURE__*/_react.default.createElement("div", {
      className: wrapperCls
    }, /*#__PURE__*/_react.default.createElement("ul", {
      className: "".concat(prefixCls, "-items")
    }, inner));
  }

  render() {
    const time = this.renderTime();
    return time;
  }

}

exports.default = timeCol;
timeCol.propTypes = {
  className: _propTypes.default.string,
  renderTimeDisplay: _propTypes.default.func
};
timeCol.contextType = _context.default;
