// @douyinfe/semi-ui/lib/cjs/calendar/weekCalendar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _foundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/calendar/foundation"));

var _localeConsumer = _interopRequireDefault(require("../locale/localeConsumer"));

var _context = _interopRequireDefault(require("../locale/context"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/calendar/constants");

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _dayCol = _interopRequireDefault(require("./dayCol"));

var _timeCol = _interopRequireDefault(require("./timeCol"));

var _eventUtil = require("@douyinfe/semi-foundation/lib/cjs/calendar/eventUtil");

require("@douyinfe/semi-foundation/lib/cjs/calendar/calendar.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const toPercent = num => {
  const res = num < 1 ? num * 100 : 100;
  return "".concat(res, "%");
};

const prefixCls = "".concat(_constants.cssClasses.PREFIX, "-week");
const allDayCls = "".concat(_constants.cssClasses.PREFIX, "-all-day");

class WeekCalendar extends _baseComponent.default {
  constructor(props) {
    super(props);

    this.checkWeekend = val => this.foundation.checkWeekend(val);

    this.handleClick = (e, val) => {
      const {
        onClick
      } = this.props;
      const value = this.foundation.formatCbValue(val);
      onClick && onClick(e, value);
    };

    this.renderDayGrid = () => {
      const {
        parsedEvents
      } = this.state;
      const events = parsedEvents.day;
      const {
        week
      } = this.weeklyData;
      const {
        markWeekend,
        dateGridRender
      } = this.props;
      const inner = week.map(day => {
        const dateString = day.date.toString();
        const dayEvents = events.has(dateString) ? events.get(dateString) : [];
        const parsed = this.foundation.getParseDailyEvents(dayEvents, day.date);
        return /*#__PURE__*/_react.default.createElement(_dayCol.default, {
          key: "".concat(dateString, "-weekday"),
          displayValue: day.date,
          scrollHeight: this.state.scrollHeight,
          handleClick: this.handleClick,
          events: parsed.day,
          showCurrTime: this.props.showCurrTime,
          isWeekend: markWeekend && day.isWeekend,
          dateGridRender: dateGridRender
        });
      });
      return inner;
    };

    this.renderHeader = dateFnsLocale => {
      const {
        markWeekend,
        displayValue
      } = this.props;
      const {
        month,
        week
      } = this.foundation.getWeeklyData(displayValue, dateFnsLocale);
      return /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "-header")
      }, /*#__PURE__*/_react.default.createElement("ul", {
        className: "".concat(_constants.cssClasses.PREFIX, "-tag ").concat(prefixCls, "-tag ").concat(prefixCls, "-sticky-left")
      }, /*#__PURE__*/_react.default.createElement("span", null, month)), /*#__PURE__*/_react.default.createElement("div", {
        role: "gridcell",
        className: "".concat(prefixCls, "-grid")
      }, /*#__PURE__*/_react.default.createElement("ul", {
        className: "".concat(prefixCls, "-grid-row")
      }, week.map(day => {
        const {
          date,
          dayString,
          weekday,
          isToday
        } = day;
        const listCls = (0, _classnames.default)({
          ["".concat(_constants.cssClasses.PREFIX, "-today")]: isToday,
          ["".concat(_constants.cssClasses.PREFIX, "-weekend")]: markWeekend && day.isWeekend
        });
        return /*#__PURE__*/_react.default.createElement("li", {
          key: "".concat(date.toString(), "-weekheader"),
          className: listCls
        }, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(_constants.cssClasses.PREFIX, "-today-date")
        }, dayString), /*#__PURE__*/_react.default.createElement("span", null, weekday));
      }))));
    };

    this.renderAllDayEvents = events => {
      const list = events.map((event, ind) => {
        const {
          leftPos,
          width,
          topInd,
          children,
          key
        } = event;
        const top = "".concat(topInd, "em");
        const style = {
          left: toPercent(leftPos),
          width: toPercent(width),
          top
        };
        return /*#__PURE__*/_react.default.createElement("li", {
          className: "".concat(_constants.cssClasses.PREFIX, "-event-item ").concat(_constants.cssClasses.PREFIX, "-event-allday"),
          key: "allDay-".concat(ind),
          style: style
        }, children);
      });
      return list;
    };

    this.renderAllDay = locale => {
      const {
        allDay
      } = this.state.parsedEvents;
      const parsed = this.foundation.parseWeeklyAllDayEvents(allDay);
      const maxRowHeight = (0, _eventUtil.calcRowHeight)(parsed);
      const style = {
        height: "".concat(maxRowHeight, "em")
      };
      const {
        markWeekend
      } = this.props;
      const {
        week
      } = this.weeklyData;
      return /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(allDayCls),
        style: style
      }, /*#__PURE__*/_react.default.createElement("ul", {
        className: "".concat(_constants.cssClasses.PREFIX, "-tag ").concat(allDayCls, "-tag ").concat(prefixCls, "-sticky-left")
      }, /*#__PURE__*/_react.default.createElement("span", null, locale.allDay)), /*#__PURE__*/_react.default.createElement("div", {
        role: "gridcell",
        className: "".concat(_constants.cssClasses.PREFIX, "-content ").concat(allDayCls, "-content")
      }, /*#__PURE__*/_react.default.createElement("ul", {
        className: "".concat(allDayCls, "-skeleton")
      }, Object.keys(week).map((date, ind) => {
        const listCls = (0, _classnames.default)({
          ["".concat(_constants.cssClasses.PREFIX, "-weekend")]: markWeekend && week[date].isWeekend
        });
        return /*#__PURE__*/_react.default.createElement("li", {
          key: "".concat(date, "-weekgrid"),
          className: listCls
        });
      })), /*#__PURE__*/_react.default.createElement("ul", {
        className: "".concat(_constants.cssClasses.PREFIX, "-event-items")
      }, this.renderAllDayEvents(parsed))));
    };

    this.state = {
      scrollHeight: 0,
      parsedEvents: {
        day: new Map(),
        allDay: new Map()
      },
      cachedKeys: []
    };
    this.foundation = new _foundation.default(this.adapter);
    this.dom = /*#__PURE__*/_react.default.createRef();
    this.scrollDom = /*#__PURE__*/_react.default.createRef();
    this.handleClick = this.handleClick.bind(this);
    this.allDayRowHeight = 1;
  }

  get adapter() {
    return Object.assign(Object.assign({}, super.adapter), {
      setWeeklyData: data => {
        this.weeklyData = data;
      },
      getWeeklyData: () => this.weeklyData,
      updateScrollHeight: scrollHeight => {
        this.setState({
          scrollHeight
        });
      },
      setParsedEvents: parsedEvents => {
        this.setState({
          parsedEvents: parsedEvents
        });
      },
      cacheEventKeys: cachedKeys => {
        this.setState({
          cachedKeys
        });
      }
    });
  }

  componentDidMount() {
    this.foundation.init();
    const {
      scrollHeight
    } = this.scrollDom.current;
    this.dom.current.scrollTop = this.props.scrollTop;
    this.foundation.notifyScrollHeight(scrollHeight);
    this.foundation.parseWeeklyEvents();
  }

  componentDidUpdate(prevProps, prevState) {
    const prevEventKeys = prevState.cachedKeys;
    const nowEventKeys = this.props.events.map(event => event.key);

    if (!(0, _isEqual2.default)(prevEventKeys, nowEventKeys)) {
      this.foundation.parseWeeklyEvents();
    }
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  render() {
    const {
      renderTimeDisplay,
      className,
      height,
      width,
      style,
      header
    } = this.props;
    const weekCls = (0, _classnames.default)(prefixCls, className);
    const weekStyle = Object.assign({
      height,
      width
    }, style);
    return /*#__PURE__*/_react.default.createElement(_localeConsumer.default, {
      componentName: "Calendar"
    }, (locale, localeCode, dateFnsLocale) => /*#__PURE__*/_react.default.createElement("div", {
      className: weekCls,
      style: weekStyle,
      ref: this.dom
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-sticky-top")
    }, header, this.renderHeader(dateFnsLocale), this.renderAllDay(locale)), /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-scroll-wrapper")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-scroll"),
      ref: this.scrollDom
    }, /*#__PURE__*/_react.default.createElement(_timeCol.default, {
      className: "".concat(prefixCls, "-sticky-left"),
      renderTimeDisplay: renderTimeDisplay
    }), this.renderDayGrid()))));
  }

}

exports.default = WeekCalendar;
WeekCalendar.propTypes = {
  displayValue: _propTypes.default.instanceOf(Date),
  header: _propTypes.default.node,
  events: _propTypes.default.array,
  mode: _propTypes.default.string,
  showCurrTime: _propTypes.default.bool,
  markWeekend: _propTypes.default.bool,
  scrollTop: _propTypes.default.number,
  renderTimeDisplay: _propTypes.default.func,
  dateGridRender: _propTypes.default.func,
  width: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  height: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
WeekCalendar.defaultProps = {
  displayValue: new Date(),
  events: [],
  mode: 'week'
};
WeekCalendar.contextType = _context.default;
