// @douyinfe/semi-ui/lib/cjs/skeleton/item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Title = exports.Paragraph = exports.Image = exports.Button = exports.Avatar = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/skeleton/constants");

var _constants2 = require("@douyinfe/semi-foundation/lib/cjs/avatar/constants");

require("@douyinfe/semi-foundation/lib/cjs/skeleton/skeleton.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const sizeSet = _constants2.strings.SIZE;
const shapeSet = _constants2.strings.SHAPE;

const generator = type => BasicComponent => props => /*#__PURE__*/_react.default.createElement(BasicComponent, Object.assign({
  type: type
}, props));

class Generic extends _react.PureComponent {
  render() {
    const _a = this.props,
          {
      prefixCls,
      className,
      type,
      size,
      shape
    } = _a,
          others = __rest(_a, ["prefixCls", "className", "type", "size", "shape"]);

    const classString = (0, _classnames.default)(className, "".concat(prefixCls, "-").concat(type), {
      ["".concat(prefixCls, "-").concat(type, "-").concat(size)]: type.toUpperCase() === 'AVATAR'
    }, {
      ["".concat(prefixCls, "-").concat(type, "-").concat(shape)]: type.toUpperCase() === 'AVATAR'
    });
    return /*#__PURE__*/_react.default.createElement('div', Object.assign({
      className: classString
    }, others));
  }

}

Generic.propTypes = {
  type: _propTypes.default.string,
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  size: _propTypes.default.oneOf(sizeSet),
  shape: _propTypes.default.oneOf(shapeSet)
};
Generic.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX,
  size: 'medium',
  shape: 'circle'
};
const Avatar = generator('avatar')(Generic);
exports.Avatar = Avatar;
const Image = generator('image')(Generic);
exports.Image = Image;
const Title = generator('title')(Generic);
exports.Title = Title;
const Button = generator('button')(Generic);
exports.Button = Button;

class Paragraph extends _react.PureComponent {
  render() {
    const {
      prefixCls,
      className,
      style,
      rows
    } = this.props;
    const classString = (0, _classnames.default)(className, "".concat(prefixCls, "-paragraph"));
    return /*#__PURE__*/_react.default.createElement("ul", {
      className: classString,
      style: style
    }, [...Array(rows)].map((e, i) => /*#__PURE__*/_react.default.createElement("li", {
      key: i
    })));
  }

}

exports.Paragraph = Paragraph;
Paragraph.propTypes = {
  rows: _propTypes.default.number,
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
Paragraph.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX,
  rows: 4
};
