// @douyinfe/semi-ui/lib/cjs/timeline/item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/timeline/constants");

require("@douyinfe/semi-foundation/lib/cjs/timeline/timeline.css");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const prefixCls = _constants.cssClasses.ITEM;

class Item extends _react.PureComponent {
  render() {
    const {
      className,
      color,
      children,
      dot,
      type,
      style,
      time,
      extra,
      onClick
    } = this.props;
    const itemCls = (0, _classnames.default)(prefixCls, className);
    const dotCls = (0, _classnames.default)({
      ["".concat(prefixCls, "-head")]: true,
      ["".concat(prefixCls, "-head-custom")]: dot,
      ["".concat(prefixCls, "-head-").concat(type)]: type
    });
    const dotStyle = color ? {
      style: {
        backgroundColor: color
      }
    } : null;
    return /*#__PURE__*/_react.default.createElement("li", {
      className: itemCls,
      style: style,
      onClick: onClick
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-tail"),
      "aria-hidden": true
    }), /*#__PURE__*/_react.default.createElement("div", Object.assign({
      className: dotCls,
      "aria-hidden": true
    }, dotStyle), dot), /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-content")
    }, children, extra && /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-content-extra")
    }, extra), time && /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-content-time")
    }, time)));
  }

}

exports.default = Item;
Item.propTypes = {
  color: _propTypes.default.string,
  time: _propTypes.default.node,
  type: _propTypes.default.oneOf(_constants.strings.ITEM_TYPE),
  dot: _propTypes.default.node,
  extra: _propTypes.default.node,
  position: _propTypes.default.oneOf(_constants.strings.ITEM_POS),
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  onClick: _propTypes.default.func
};
Item.defaultProps = {
  type: 'default',
  time: '',
  onClick: _noop2.default
};
