// @douyinfe/semi-ui/lib/cjs/steps/navSteps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/steps/constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const Steps = props => {
  const {
    size,
    current,
    initial,
    children,
    prefixCls,
    className,
    style,
    onChange
  } = props;
  const inner = (0, _react.useMemo)(() => {
    const filteredChildren = _react.Children.toArray(children).filter(c => /*#__PURE__*/(0, _react.isValidElement)(c));

    const total = filteredChildren.length;

    const content = _react.Children.map(filteredChildren, (child, index) => {
      if (!child) {
        return null;
      }

      const childProps = Object.assign({
        index,
        total
      }, child.props);
      childProps.active = index === current;

      childProps.onChange = () => {
        if (index !== current) {
          onChange(index + initial);
        }
      };

      return /*#__PURE__*/(0, _react.cloneElement)(child, Object.assign({}, childProps));
    });

    return content;
  }, [children, prefixCls, current, size, initial, onChange]);
  const wrapperCls = (0, _classnames.default)(className, {
    ["".concat(prefixCls, "-nav")]: true,
    ["".concat(prefixCls, "-").concat(size)]: size !== 'default'
  });
  return /*#__PURE__*/_react.default.createElement("div", {
    "aria-label": props["aria-label"],
    className: wrapperCls,
    style: style
  }, inner);
};

Steps.propTypes = {
  prefixCls: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  current: _propTypes.default.number,
  initial: _propTypes.default.number,
  size: _propTypes.default.oneOf(['small', 'default'])
};
Steps.defaultProps = {
  prefixCls: _constants.stepsClasses.PREFIX,
  current: 0,
  direction: 'horizontal',
  size: 'default',
  initial: 0,
  status: 'process'
};
var _default = Steps;
exports.default = _default;
