// @douyinfe/semi-ui/lib/es/image/previewFooter.js
import _throttle from "lodash/throttle";
import React from "react";
import BaseComponent from "../_base/baseComponent";
import { IconChevronLeft, IconChevronRight, IconMinus, IconPlus, IconRotate, IconDownload, IconWindowAdaptionStroked, IconRealSizeStroked } from "@douyinfe/semi-icons";
import PropTypes from "prop-types";
import Tooltip from "../tooltip";
import Divider from "../divider";
import Slider from "../slider";
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/image/constants';
import cls from "classnames";
import PreviewFooterFoundation from '@douyinfe/semi-foundation/lib/es/image/previewFooterFoundation';
import LocaleConsumer from "../locale/localeConsumer";
const prefixCls = cssClasses.PREFIX;
const footerPrefixCls = "".concat(cssClasses.PREFIX, "-preview-footer");
let mouseActiveTime = 0;
export default class Footer extends BaseComponent {
  constructor(props) {
    super(props);

    this.changeSliderValue = type => {
      this.foundation.changeSliderValue(type);
    };

    this.handleMinusClick = () => {
      this.changeSliderValue("minus");
    };

    this.handlePlusClick = () => {
      this.changeSliderValue("plus");
    };

    this.handleRotateLeft = () => {
      this.foundation.handleRotate("left");
    };

    this.handleRotateRight = () => {
      this.foundation.handleRotate("right");
    };

    this.handleSlideChange = _throttle(value => {
      this.foundation.handleValueChange(value);
    }, 50);

    this.handleRatioClick = () => {
      this.foundation.handleRatioClick();
    };

    this.customRenderViewMenu = () => {
      const {
        min,
        max,
        step,
        curPage,
        totalNum,
        ratio,
        zoom,
        disabledPrev,
        disabledNext,
        disableDownload,
        onNext,
        onPrev,
        onDownload,
        renderPreviewMenu
      } = this.props;
      const props = {
        min,
        max,
        step,
        curPage,
        totalNum,
        ratio,
        zoom,
        disabledPrev,
        disabledNext,
        disableDownload,
        onNext,
        onPrev,
        onDownload,
        onRotateLeft: this.handleRotateLeft,
        onRotateRight: this.handleRotateRight,
        disabledZoomIn: zoom === max,
        disabledZoomOut: zoom === min,
        onRatioClick: this.handleRatioClick,
        onZoomIn: this.handlePlusClick,
        onZoomOut: this.handleMinusClick
      };
      return renderPreviewMenu(props);
    }; // According to showTooltip in props, decide whether to use Tooltip to pack a layer
    // 根据 props 中的 showTooltip 决定是否使用 Tooltip 包一层


    this.getFinalIconElement = (element, content) => {
      const {
        showTooltip
      } = this.props;
      return showTooltip ? /*#__PURE__*/React.createElement(Tooltip, {
        content: content
      }, element) : element;
    };

    this.getLocalTextByKey = key => /*#__PURE__*/React.createElement(LocaleConsumer, {
      componentName: "Image"
    }, locale => locale[key]);

    this.getIconChevronLeft = () => {
      const {
        disabledPrev,
        onPrev,
        prevTip
      } = this.props;
      const icon = /*#__PURE__*/React.createElement(IconChevronLeft, {
        size: "large",
        className: disabledPrev ? "".concat(footerPrefixCls, "-disabled") : "",
        onClick: !disabledPrev ? onPrev : undefined
      });
      const content = prevTip !== null && prevTip !== void 0 ? prevTip : this.getLocalTextByKey("prevTip");
      return this.getFinalIconElement(icon, content);
    };

    this.getIconChevronRight = () => {
      const {
        disabledNext,
        onNext,
        nextTip
      } = this.props;
      const icon = /*#__PURE__*/React.createElement(IconChevronRight, {
        size: "large",
        className: disabledNext ? "".concat(footerPrefixCls, "-disabled") : "",
        onClick: !disabledNext ? onNext : undefined
      });
      const content = nextTip !== null && nextTip !== void 0 ? nextTip : this.getLocalTextByKey("nextTip");
      return this.getFinalIconElement(icon, content);
    };

    this.getIconMinus = () => {
      const {
        zoomOutTip,
        zoom,
        min
      } = this.props;
      const disabledZoomOut = zoom === min;
      const icon = /*#__PURE__*/React.createElement(IconMinus, {
        size: "large",
        onClick: !disabledZoomOut ? this.handleMinusClick : undefined,
        className: disabledZoomOut ? "".concat(footerPrefixCls, "-disabled") : ""
      });
      const content = zoomOutTip !== null && zoomOutTip !== void 0 ? zoomOutTip : this.getLocalTextByKey("zoomOutTip");
      return this.getFinalIconElement(icon, content);
    };

    this.getIconPlus = () => {
      const {
        zoomInTip,
        zoom,
        max
      } = this.props;
      const disabledZoomIn = zoom === max;
      const icon = /*#__PURE__*/React.createElement(IconPlus, {
        size: "large",
        onClick: !disabledZoomIn ? this.handlePlusClick : undefined,
        className: disabledZoomIn ? "".concat(footerPrefixCls, "-disabled") : ""
      });
      const content = zoomInTip !== null && zoomInTip !== void 0 ? zoomInTip : this.getLocalTextByKey("zoomInTip");
      return this.getFinalIconElement(icon, content);
    };

    this.getIconRatio = () => {
      const {
        ratio,
        originTip,
        adaptiveTip
      } = this.props;
      const props = {
        size: "large",
        className: cls("".concat(footerPrefixCls, "-gap")),
        onClick: this.handleRatioClick
      };
      const icon = ratio === "adaptation" ? /*#__PURE__*/React.createElement(IconRealSizeStroked, Object.assign({}, props)) : /*#__PURE__*/React.createElement(IconWindowAdaptionStroked, Object.assign({}, props));
      let content;

      if (ratio === "adaptation") {
        content = originTip !== null && originTip !== void 0 ? originTip : this.getLocalTextByKey("originTip");
      } else {
        content = adaptiveTip !== null && adaptiveTip !== void 0 ? adaptiveTip : this.getLocalTextByKey("adaptiveTip");
      }

      return this.getFinalIconElement(icon, content);
    };

    this.getIconRotate = () => {
      const {
        rotateTip
      } = this.props;
      const icon = /*#__PURE__*/React.createElement(IconRotate, {
        size: "large",
        onClick: this.handleRotateLeft
      });
      const content = rotateTip !== null && rotateTip !== void 0 ? rotateTip : this.getLocalTextByKey("rotateTip");
      return this.getFinalIconElement(icon, content);
    };

    this.getIconDownload = () => {
      const {
        downloadTip,
        onDownload,
        disableDownload
      } = this.props;
      const icon = /*#__PURE__*/React.createElement(IconDownload, {
        size: "large",
        onClick: !disableDownload ? onDownload : undefined,
        className: cls("".concat(footerPrefixCls, "-gap"), {
          ["".concat(footerPrefixCls, "-disabled")]: disableDownload
        })
      });
      const content = downloadTip !== null && downloadTip !== void 0 ? downloadTip : this.getLocalTextByKey("downloadTip");
      return this.getFinalIconElement(icon, content);
    };

    this.foundation = new PreviewFooterFoundation(this.adapter);
  }

  get adapter() {
    return Object.assign(Object.assign({}, super.adapter), {
      setStartMouseOffset: time => {
        mouseActiveTime = time;
      }
    });
  }

  render() {
    const {
      min,
      max,
      step,
      curPage,
      totalNum,
      zoom,
      showTooltip,
      className,
      renderPreviewMenu
    } = this.props;

    if (renderPreviewMenu) {
      return /*#__PURE__*/React.createElement("div", {
        className: "".concat(footerPrefixCls, "-wrapper")
      }, this.customRenderViewMenu());
    }

    return /*#__PURE__*/React.createElement("section", {
      className: cls(footerPrefixCls, "".concat(footerPrefixCls, "-wrapper"), className)
    }, this.getIconChevronLeft(), /*#__PURE__*/React.createElement("div", {
      className: "".concat(footerPrefixCls, "-page")
    }, /*#__PURE__*/React.createElement("span", null, curPage), /*#__PURE__*/React.createElement("span", null, "/"), /*#__PURE__*/React.createElement("span", null, totalNum)), this.getIconChevronRight(), /*#__PURE__*/React.createElement(Divider, {
      layout: "vertical"
    }), this.getIconMinus(), /*#__PURE__*/React.createElement(Slider, {
      value: zoom,
      min: min,
      max: max,
      step: step,
      tipFormatter: v => "".concat(v, "%"),
      tooltipVisible: showTooltip ? undefined : false,
      onChange: this.handleSlideChange
    }), this.getIconPlus(), this.getIconRatio(), /*#__PURE__*/React.createElement(Divider, {
      layout: "vertical"
    }), this.getIconRotate(), this.getIconDownload());
  }

}
Footer.propTypes = {
  curPage: PropTypes.number,
  totalNum: PropTypes.number,
  disabledPrev: PropTypes.bool,
  disabledNext: PropTypes.bool,
  disableDownload: PropTypes.bool,
  className: PropTypes.string,
  zoom: PropTypes.number,
  ratio: PropTypes.string,
  prevTip: PropTypes.string,
  nextTip: PropTypes.string,
  zoomInTip: PropTypes.string,
  zoomOutTip: PropTypes.string,
  rotateTip: PropTypes.string,
  downloadTip: PropTypes.string,
  adaptiveTip: PropTypes.string,
  originTip: PropTypes.string,
  showTooltip: PropTypes.bool,
  onZoomIn: PropTypes.func,
  onZoomOut: PropTypes.func,
  onPrev: PropTypes.func,
  onNext: PropTypes.func,
  onAdjustRatio: PropTypes.func,
  onRotateLeft: PropTypes.func,
  onDownload: PropTypes.func
};
Footer.defaultProps = {
  min: 10,
  max: 500,
  step: 10,
  showTooltip: false,
  disableDownload: false
};
